### PR TITLE
feat(roll): roll to Playwright 1.12.0-next-1621456974000

### PR DESCRIPTION
**Pull request by kumaraditya303**
Roll Playwright driver to 1.12.0-next-1620880294000

### CODE REVIEW

File: playwright/_impl/_transport.py [hunk 1]
```diff
@@ -153,7 +153,6 @@ def __init__(
     ) -> None:
         super().__init__(loop)
         Transport.__init__(self, loop)
-
```
Comment (mxschmitt) [inline playwright/_impl/_transport.py:156]: please revert this, since there is no actual change in this file.

File: playwright/_impl/_browser.py [hunk 2]
```diff
@@ -148,6 +148,12 @@ async def new_page(
     async def close(self) -> None:
         if self._is_closed_or_closing:
             return
+        if self._is_remote:
```
Comment (mxschmitt) [inline playwright/_impl/_browser.py:151]: It's possible that the WebSocket connection get's disconnected without the user's control. This will end up that the Browser.close method get's never called. Moving the loop to when the WebSocket connection actually closes would fix it, see here how it's done upstream:

https://github.com/microsoft/playwright/blob/e018c5dd426c7cef9bbf6eb497112083013bb6ba/src/client/browserType.ts#L170-L179

Seems like there is no test yet for it, having one for that would be neat.

**Comment by mxschmitt** [comment]
Please make the file change from CRLF to LF in another PR.

File: playwright/_impl/_browser.py (hunk 2 above)
Comment (mxschmitt) [reply playwright/_impl/_browser.py:156]: Upstream we have this loop only a single time, would be good if we could have it here also at a single place only. It feels wrong to have it inside: _on_close.

Comment (kumaraditya303) [reply playwright/_impl/_browser.py:156]: When the transport closes it calls `_on_close` when websocket connection is closed

File: tests/async/test_tracing.py [hunk 3]
```diff
@@ -103,3 +103,18 @@ async def test_should_support_a_buffer_without_a_path(
     await page.goto(server.PREFIX + "/grid.html")
     trace = await browser.stop_tracing()
     assert "screenshot" in trace.decode()
+
+
+@pytest.mark.only_browser("chromium")
```
Comment (mxschmitt) [inline tests/async/test_tracing.py:108]: this is different from the existing tracing, so it's not Chromium specific. Also it might make sense to move the existing tracing into a `test_chromium_tracing.py` and the new tracing into `test_tracing.py`

File: playwright/_impl/_browser.py (hunk 2 above)
Comment (mxschmitt) [reply playwright/_impl/_browser.py:156]: Upstream we do it next to the ws close event and not in the browser, I would move it into our browser_type.py when we receive the disconnect event from the WS client.

See here how it's done upstream: https://github.com/microsoft/playwright/blob/4c3bd11820665274071ee24663a58594a0b34b94/src/client/browserType.ts#L168-L177

Comment (kumaraditya303) [reply playwright/_impl/_browser.py:156]: It uses event listener for websocket close but in python we don't have that, for that a custom protocol class would be required which on close would do that

File: playwright/_impl/_browser.py [hunk 4]
```diff
@@ -60,6 +60,11 @@ def __repr__(self) -> str:
 
     def _on_close(self) -> None:
         self._is_connected = False
+        if self._is_remote:
```
Comment (kumaraditya303) [inline playwright/_impl/_browser.py:63]: This is currently used when the transport closes unexpectedly

File: playwright/_impl/_browser.py [hunk 5]
```diff
@@ -148,6 +153,12 @@ async def new_page(
     async def close(self) -> None:
         if self._is_closed_or_closing:
             return
+        if self._is_remote:
```
Comment (kumaraditya303) [inline playwright/_impl/_browser.py:156]: This is used when user calls close

File: playwright/_impl/_browser.py (hunk 2 above)
Comment (mxschmitt) [reply playwright/_impl/_browser.py:156]: We have it, see here :) https://github.com/microsoft/playwright-python/blob/master/playwright/_impl/_browser_type.py#L212

Comment (mxschmitt) [reply playwright/_impl/_browser.py:151]: In my opinion we still should move this code away from the browser and into connect method as it was done upstream since it's specific to remote connections via connect.

**Review by mxschmitt — APPROVED** [review]

File: playwright/_impl/_browser_context.py [hunk 6]
```diff
@@ -299,3 +372,7 @@ async def new_cdp_session(self, page: Page) -> CDPSession:
         return from_channel(
             await self._channel.send("newCDPSession", {"page": page._channel})
         )
+
+    @property
+    def tracing(self) -> Tracing:
```
Comment (dgozman) [inline playwright/_impl/_browser_context.py:377]: I don't think we are ready for the tracing just yet, we are still working on the API shape. Can we roll without tracing?

File: playwright/_impl/_browser_type.py [hunk 7]
```diff
@@ -209,7 +212,14 @@ async def connect(
         browser._is_remote = True
         browser._is_connected_over_websocket = True
 
-        transport.once("close", browser._on_close)
+        def handle_transport_close() -> None:
```
Comment (dgozman) [inline playwright/_impl/_browser_type.py:215]: Why is this a part of the roll? Is it a bug fix?

File: tests/async/test_tracing.py (hunk 3 above)
Comment (dgozman) [reply tests/async/test_tracing.py:22]: Please drop the marker, as Max says.

File: tests/sync/test_browsertype_connect.py [hunk 8]
```diff
@@ -26,10 +26,10 @@ def test_browser_type_connect_slow_mo(
     remote_server = launch_server()
     browser = browser_type.connect(remote_server.ws_endpoint, slow_mo=100)
     browser_context = browser.new_context()
-    page = browser_context.new_page()
     t1 = time.monotonic()
+    page = browser_context.new_page()
     assert page.evaluate("11 * 11") == 121
-    assert (time.monotonic() - t1) >= 0.100
+    assert (time.monotonic() - t1) >= 0.2
```
Comment (dgozman) [inline tests/sync/test_browsertype_connect.py:32]: Why this change?

File: playwright/_impl/_transport.py [hunk 9]
```diff
@@ -161,6 +161,7 @@ def __init__(
 
     def request_stop(self) -> None:
         self._stopped = True
+        self.emit("close")
```
Comment (dgozman) [inline playwright/_impl/_transport.py:164]: I did not find a place that listens to this. Is it a bug fix?

File: playwright/_impl/_browser_type.py (hunk 7 above)
Comment (mxschmitt) [reply playwright/_impl/_browser_type.py:215]: The disconnect handling was changed from you in the last roll, I adopted your changes.

File: tests/sync/test_browsertype_connect.py (hunk 8 above)
Comment (mxschmitt) [reply tests/sync/test_browsertype_connect.py:32]: it was flaky before, this should "fix" the flakyness

File: playwright/_impl/_transport.py (hunk 9 above)
Comment (mxschmitt) [reply playwright/_impl/_transport.py:164]: with the new disconnect handling, we use it to listen on the transport websocket disconnect via this

File: tests/sync/test_browsertype_connect.py (hunk 8 above)
Comment (mxschmitt) [reply tests/sync/test_browsertype_connect.py:32]: extracted in https://github.com/microsoft/playwright-python/pull/709

File: playwright/_impl/_browser_context.py (hunk 6 above)
Comment (mxschmitt) [reply playwright/_impl/_browser_context.py:377]: discussed offline, rolling with it and maybe reverting / changing it then.

Comment (kumaraditya303) [reply playwright/_impl/_browser_context.py:377]: @dgozman @mxschmitt fyi: Tracing was implemented in playwright-java too https://github.com/microsoft/playwright-java/pull/444

File: playwright/_impl/_transport.py (hunk 9 above)
Comment (kumaraditya303) [reply playwright/_impl/_transport.py:164]: Required for when the websocket connection is closed without client closing

File: tests/sync/test_browsertype_connect.py (hunk 8 above)
Comment (kumaraditya303) [reply tests/sync/test_browsertype_connect.py:32]: Fine to extract it away but should allow these kind of small fixes for flaky tests anyway
